src/signalrclient/signalr_default_scheduler.cpp

```diff
@@ -0,0 +1,217 @@
+                }
+
+                assert(internals->m_callback == nullptr);
+            });
+    }
+
+    void thread::add(signalr_base_cb cb)
+    {
+        {
+            assert(m_internals->m_closed == false);
+            assert(m_internals->m_busy == false);
+
+            std::lock_guard<std::mutex> lock(m_internals->m_callback_lock);
+            m_internals->m_callback = cb;
+            m_internals->m_busy = true;
+        } // unlock
+        m_internals->m_callback_cv.notify_one();
+    }
+
+    void thread::shutdown()
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_internals->m_callback_lock);
+            m_internals->m_closed = true;
+        }
+        m_internals->m_callback_cv.notify_one();
+        m_thread.join();
+    }
+
+    bool thread::is_free() const
+    {
+        return !m_internals->m_busy;
+    }
+
+    thread::~thread()
+    {
+        shutdown();
+    }
+
+    void signalr_default_scheduler::schedule(const signalr_base_cb& cb, std::chrono::milliseconds delay)
+    {
+        {
+            assert(m_internals->m_closed == false);
+            std::lock_guard<std::mutex> lock(m_internals->m_callback_lock);
+            m_internals->m_callbacks.push_back(std::make_pair(cb, std::chrono::steady_clock::now() + delay));
+        } // unlock
+
+        // If callback has a delay, no point in running the logic to check if it should run early
+        if (delay == std::chrono::milliseconds::zero())
+        {
+            m_internals->m_callback_cv.notify_one();
+        }
+    }
+
+    void signalr_default_scheduler::run()
+    {
+        auto internals = m_internals;
+
+        std::thread([=]()
+            {
+                std::vector<thread> threads{ 5 };
+
+                std::unique_lock<std::mutex> lock(internals->m_callback_lock);
+
+                auto prev_callback_count = size_t{ 0 };
+
+                while (true)
+                {
+                    auto& callbacks = internals->m_callbacks;
+                    auto& closed = internals->m_closed;
+
+                    if (closed && callbacks.empty())
+                    {
+                        assert(callbacks.empty());
+                        return;
+                    }
+
+                    internals->m_callback_cv.wait_for(lock, std::chrono::milliseconds(15),
+                        [&callbacks, &closed, &prev_callback_count]
+                        {
+                            return closed || prev_callback_count != callbacks.size();
+                        });
+
+                    // find the first callback that is ready to run, find a thread to run it and remove it from the list
+                    auto curr_time = std::chrono::steady_clock::now();
+                    auto it = callbacks.begin();
+                    auto found = false;
+                    while (it != callbacks.end())
+                    {
+                        if (it->second <= curr_time)
```

Comment:
I don't understand this time check here and what the goal is.

Reply:
If a task was scheduled, but told not to run for X seconds, we add the delay to the time (line 110) when it was added and then do the check here to see if it's ready to run.